spark_client/domain.py

```diff
@@ -176,30 +176,35 @@ def __init__(self, environ: Dict = dict(os.environ)):
         self.environ = environ if environ is not None else {}
 
     @property
-    def spark_folder(self) -> str:
+    def spark_home(self) -> str:
+        """Return SPARK_HOME folder"""
+        return self.environ.get("SPARK_HOME", os.path.dirname(__file__))
+
+    @property
+    def spark_install_base(self) -> str:
         """Return the SNAP folder"""
-        return self.environ.get("SNAP", self.environ.get("SPARK_HOME"))
+        return self.environ.get("SPARK_INSTALL_BASE", self.spark_home)
 
     @property
-    def static_conf_file(self) -> str:
-        """Return static config properties file packaged with the client snap."""
-        return f"{self.spark_folder}/conf/spark-defaults.conf"
+    def spark_user_home(self) -> str:
+        """Return SPARK_HOME folder"""
+        return self.environ.get("HOME", self.spark_install_base)
 
     @property
-    def dynamic_conf_file(self) -> str:
-        """Return dynamic config properties file generated during client setup."""
-        return f"{self.environ.get('SNAP_USER_DATA', self.environ.get('HOME'))}/spark-defaults.conf"
+    def spark_data_folder(self) -> str:
+        """Return the data folder to be used by Spark"""
+        return self.environ.get("SPARK_DATA", self.spark_home)
```

Comment:
Seemed to me that we can have the following layers of locations:

 - `SPARK_HOME` - a Spark-level home definition
 - Installation base (typically the `snap` base directory)
 - The Spark `USER` home directory
 - A Spark 'data' directory

Let me know if any of those may be redundant, and should go.

Reply:
Furthermore, let me know if I may have not understood the functionality of any of those correctly (i.e. mistakes in the references below or all around) :-) 

Though the tests are passing, it may be just by co-incidence (of all these pointers targeting the local build directory here -- could easily break when used from a snap!)

Reply:
Probably it is sensible to have these defined:

1. spark_home (this should be the root of the spark folder, spark_home is really a widely used concept in spark, in general we may assume this is READ-ONLY)
2. package_dir (directory where the package is located, this is useful for retrieving package resources, such as the templates, in general we may assume this is READ-ONLY)
3. confs folder (to have a similar meaning and segregation, as in the [spec](https://docs.google.com/document/d/1MyT4Wt_-vgPwhwEjC-oeHRdLYrZWaQzFGTwclT7hD1U/edit), in general we may assume this can be written by the user, and rather user and version specific)
4. data folder (to have a similar meaning and segregation, as in the [spec](https://docs.google.com/document/d/1MyT4Wt_-vgPwhwEjC-oeHRdLYrZWaQzFGTwclT7hD1U/edit), in general we may assume this can be written by the user and it is persistent across upgrades)

Looking at your bullet points, I believe this is the mapping

> * SPARK_HOME - a Spark-level home definition 

=> 1. 

> * Installation base (typically the snap base directory) 

=> None (I don't think we need to provide this installation base, it probably coincides with SPARK_HOME

> * The Spark USER home directory 

=> 4. 

> * A Spark 'data' directory 

=> 3. 

When in a SNAP, we should then be able to bind these to some path in the SNAP model 

@welpaolo please provide review and provide feedback on this

Reply:
So 2. may not be needed (at least that's what it seemed to me, that's why I've removed it :-) ). However one, similar concept would need to be kept around still -- at least to my limited understanding.

The original `spark_folder` (not the same as `dir_package`), that I've renamed to `spark_install_base` is typically to point to the actual installation space -- whici in our case would be `/var/snap`. I assume this may be a different location pointed by the `SPARK_HOME` environment variable, and both could (should) co-exist... (That seemed to be the logic I gathered from the code the least.) Typically when searching for the `kubectl` command: which may be located in the installation base (and if it is, we may strictly want to use that one) -- however would not be located in `SPARK_HOME`.

As for the naming terminology: should I rename according to what you've used above (I mean were those names suggestive) or it was just to identify the logic, and naming is OK as it is?

Reply:
It wasn't quite by my free choice -- the tests weren't running without adjustments of the `Defaults`.
(At which point it seemed like the clean way to do a full cleanup/re-adjustmen of the snap references on this level.)

However, for this particular reason I'm offering in the *Descriotion* of the PR to split it to a separate PR. If preferred I'm glad to do that, as it's trivial.

Reply:
Just for the record: this is now in a separate ticket with a correponding branch (starting with the code moved out from here): https://warthogs.atlassian.net/browse/DPE-1952